tests/python/pants_test/engine/test_rules.py

```diff
@@ -245,36 +246,33 @@ def test_not_fulfillable_duplicated_dependency(self):
       (D, (Select(A), SelectDependencies(A, SubA, field_types=(C,))), noop),
       (A, (Select(SubA),), noop)
     ]
-    validator = RulesetValidator(RuleIndex.create(rules, tuple()),
-      goal_to_product={},
-      root_subject_types=_suba_root_subject_types)
+    validator = self.create_validator({}, tuple(), _suba_root_subject_types, rules)
 
     with self.assertRaises(ValueError) as cm:
       validator.validate()
 
     self.assert_equal_with_printing(dedent("""
                       Rules with errors: 2
-                        (B, (Select(D),), noop):
-                          depends on unfulfillable (D, (Select(A), SelectDependencies(A, SubA, field_types=(C,))), noop) of SubA with subject types: SubA
-                        (D, (Select(A), SelectDependencies(A, SubA, field_types=(C,))), noop):
-                          depends on unfulfillable (A, (Select(SubA),), noop) of C with subject types: SubA""").strip(),
+                        (Exactly(B), (Select(Exactly(D)),), noop):
+                          depends on unfulfillable (Exactly(D), (Select(Exactly(A)), SelectDependencies(Exactly(A), Exactly(SubA), field_types=(C,))), noop) of SubA with subject types: SubA
+                        (Exactly(D), (Select(Exactly(A)), SelectDependencies(Exactly(A), Exactly(SubA), field_types=(C,))), noop):
+                          depends on unfulfillable (Exactly(A), (Select(Exactly(SubA)),), noop) of C with subject types: SubA
+                      """).strip(),
         str(cm.exception))
 
   def test_initial_select_projection_failure(self):
     rules = [
       (Exactly(A), (SelectProjection(B, D, ('some',), C),), noop),
     ]
-    validator = RulesetValidator(RuleIndex.create(rules, tuple()),
-      goal_to_product={},
-      root_subject_types=_suba_root_subject_types)
+    validator = self.create_validator({}, tuple(), _suba_root_subject_types, rules)
 
     with self.assertRaises(ValueError) as cm:
       validator.validate()
 
     self.assert_equal_with_printing(dedent("""
                       Rules with errors: 1
-                        (Exactly(A), (SelectProjection(B, D, (u'some',), C),), noop):
-                          no matches for Select(C) when resolving SelectProjection(B, D, (u'some',), C) with subject types: SubA
+                        (Exactly(A), (SelectProjection(Exactly(B), D, ("some",), Exactly(C)),), noop):
```

Comment:
I'm still really hoping that we can kill the explicit `TypeConstraint` struct in favor of a vec of `TypeId`s... ie, only support union types, and nothing else. So exposing `Exactly` like this (much more than before) seems like a step backwards.